src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp

```diff
@@ -0,0 +1,119 @@
+
+Date: March 2023
+
+\*******************************************************************/
+#include "dfcc_infer_loop_assigns.h"
+
+#include <util/expr.h>
+#include <util/find_symbols.h>
+#include <util/message.h>
+#include <util/pointer_expr.h>
+#include <util/std_code.h>
+
+#include <goto-instrument/havoc_utils.h>
+
+/// Widens expressions in the given \p assigns set as follows:
+/// If an expression is an array index expression or an object dereference
+/// expression with a non-constant offset, e.g. a[i] or *(b+i) with a
+/// non-constant `i`, then replace it by the entire underlying object.
+/// Otherwise, e.g. for a[i] or *(b+i) when `i` is a known constant,
+/// keep the expression in the result.
+/// By `constant` for the index expression, we do not mean a literal constant
+/// expression. Rather, we mean a symbol or an expression the value of which is
+/// known not to change over loop iterations, i.e. an expression that is not
+/// itself part of \p assigns.
+static void dfcc_widen_non_const(const namespacet &ns, assignst &assigns)
+{
+  assignst result;
+  const symbolt &object_whole_sym = ns.lookup(CPROVER_PREFIX "object_whole");
+  code_typet object_whole_code_type = to_code_type(object_whole_sym.type);
+  havoc_utils_is_constantt is_constant(assigns, ns);
+
+  for(const auto &e : assigns)
+  {
+    if(e.id() == ID_index || e.id() == ID_dereference)
+    {
+      address_of_exprt address_of_expr(e);
+
+      // index or offset is non-constant.
+      if(!is_constant(address_of_expr))
+      {
+        exprt::operandst args;
+        args.emplace_back(address_of_expr);
+        result.emplace(side_effect_expr_function_callt(
+          object_whole_sym.symbol_expr(),
+          args,
+          object_whole_code_type.return_type(),
+          address_of_expr.source_location()));
+      }
+      else
+        result.emplace(e);
+    }
+    else
+      result.emplace(e);
+  }
+  assigns.swap(result);
+}
+
+std::set<exprt> dfcc_infer_loop_assigns(
+  const local_may_aliast &local_may_alias,
+  const loopt &loop_instructions,
+  const source_locationt &loop_head_location,
+  messaget &log,
+  namespacet &ns)
+{
+  try
+  {
+    // infer
+    assignst assigns;
+    get_assigns(local_may_alias, loop_instructions, assigns);
+
+    // compute locals
+    std::set<irep_idt> loop_locals;
+    for(const auto &target : loop_instructions)
+    {
+      if(target->is_decl())
+      {
+        loop_locals.insert(target->decl_symbol().get_identifier());
+      }
+    }
+
+    // remove all targets that depend on loop-local identifiers
+    auto expr_it = assigns.begin();
+    while(expr_it != assigns.end())
+    {
+      const std::unordered_set<irep_idt> symbols =
+        find_symbol_identifiers(*expr_it);
+
+      if(std::find_if(symbols.begin(), symbols.end(), [&](const irep_idt &id) {
+           return loop_locals.find(id) != loop_locals.end();
+         }) != symbols.end())
+      {
+        expr_it = assigns.erase(expr_it);
+      }
```

Comment:
Is it correct to remove an expression from `assigns` if _any_ symbol occurring in there is a loop-local one? (At least that is my understanding of what the `find_if` will accomplish here.)

Reply:
Removing something from assigns is always sound as it can only cause the analysis to fail with an assigns clause violation.

We really have to remove any inferred expression that contains a non-local symbol, because assigns clause targets can only talk about variables that exist outside of the loop.

```
for (int i = 0; i < N; i++) {
 int j = i;
 arr[j] = 0;
}
```
In that case the inference returns the expressions `i` and `arr[j]` as inferred targets, and we cannot use `arr[j]`.

Maybe we could do the following instead: if a target depends on some loop local, compute `root = root_object(inferred_target)`,  if `root` still contains non-local variables, discard it, otherwise widen to `__CPROVER_object_whole(root)` and add it to the set.
 
  @qinheping does that sound right ?

Reply:
In the example, `get_assigns` will give us `j` and `arr[j]`. 
Yes, we should remove the local variable `j`.
For deference expressions with non-constant offset, e.g., `*(arr+j)`; and indexing expressions with non-constant index, e.g., `arr[j]`, we should widen them with `__CPROVER_object_whole`.

Reply:
This is what is implemented in this new revision, thanks to #7636